### PR TITLE
Bilinear in 3d

### PR DESCRIPTION
**Pull request by sac-bsa**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*
I have been working with @egpuckett on making the bi-linear least squares particle interpolator work in for three dimensions. @gassmoeller has seen a draft of this code and recommended that we make a PR. 

### For all pull requests:

* [X] I have followed the [instructions for indenting my code](../CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [X] I have tested my new feature locally to ensure it is correct.
* [X] I have created a testcase; see #3208
(http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../tests/) directory.

* [ ] I have added a changelog entry in the [doc/modules/changes](../doc/modules/changes) directory that will inform other users of my change.

Do I have to make a changelog entry for this change? And if so, please instruct.

### CODE REVIEW

File: tests/particle_interpolator_bilinear_3d.prm [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# This is a variation of the particle_interpolator_cell_average_2
+# test that includes a time varying solution with active particles.
+# In particular this test was introduced to check for the accuracy
+# of the interpolation in time-varying flow.
```
Comment (gassmoeller) [inline tests/particle_interpolator_bilinear_3d.prm]:
```suggestion
# of the bilinear least squares interpolator in 3D.
```

File: include/aspect/particle/interpolator/bilinear_least_squares.h [hunk 2]
```diff
@@ -81,6 +81,9 @@ namespace aspect
            */
           std::vector<double> global_maximum_particle_properties;
           std::vector<double> global_minimum_particle_properties;
+
+
+          bool use_neighbors;
```
Comment (gassmoeller) [inline include/aspect/particle/interpolator/bilinear_least_squares.h]: I do not see where this is used. Please remove from this PR.

**Comment by egpuckett** [comment]
@gassmoeller When you type /rebuild and the checks fail, is there anything Mack or we should do about this?

**Comment by gassmoeller** [comment]
Yes, you can download the changes the tester found (see my last post), commit them, git push again, and the tester will rerun.

**Comment by gassmoeller** [comment]
@sac-bsa is this ready for another review? Please always leave a comment if you want me to take another look, because otherwise I will not know if this is work in progress or if I should spend time looking at it (from the changes it looks like it is ready?)

**Comment by sac-bsa** [comment]
Yes, I believe that I have had the suggested changes, and modified the test so that it passed the tester. I believe that this pull request is ready for a review

File: source/particle/interpolator/bilinear_least_squares.cc [hunk 3]
```diff
@@ -110,11 +107,21 @@ namespace aspect
...
+            if (dim == 2)
+              {
+                A(index, 3) = relative_particle_position[0] * relative_particle_position[1];
+              }
+            else
+              {
+                A(index, 3) = relative_particle_position[2];
+                A(index, 4) = relative_particle_position[0] * relative_particle_position[1];
+                A(index, 5) = relative_particle_position[0] * relative_particle_position[2];
+                A(index, 6) = relative_particle_position[1] * relative_particle_position[2];
```
Comment (gassmoeller) [inline source/particle/interpolator/bilinear_least_squares.cc]: I should have noticed this earlier, but what about the shape function index 7, i.e. `x*y*z`? Should that not be here as well?

File: source/particle/interpolator/bilinear_least_squares.cc [hunk 4]
```diff
@@ -97,7 +94,7 @@ namespace aspect
         // Noticed that the size of matrix A is n_particles x matrix_dimension
         // which usually is not a square matrix. Therefore, we solve Ax=r by
         // solving A^TAx= A^Tr.
-        const unsigned int matrix_dimension = 4;
+        const unsigned int matrix_dimension = (dim == 2) ? 4: 7;
```
Comment (gassmoeller) [inline source/particle/interpolator/bilinear_least_squares.cc]: See below, but should this not be 4 : 8?

File: source/particle/interpolator/bilinear_least_squares.cc [hunk 5]
```diff
@@ -135,13 +142,23 @@ namespace aspect
...
+            if (dim == 2)
+              {
+                interpolated_value += c[3] * relative_support_point_location[0] * relative_support_point_location[1];
+              }
+            else
+              {
+                interpolated_value += c[3] * relative_support_point_location[2] +
+                                      c[4] * relative_support_point_location[0] * relative_support_point_location[1] +
+                                      c[5] * relative_support_point_location[0] * relative_support_point_location[2] +
+                                      c[6] * relative_support_point_location[1] * relative_support_point_location[2];
```
Comment (gassmoeller) [inline source/particle/interpolator/bilinear_least_squares.cc]: same here as above

**Comment by sac-bsa** [comment]
@gassmoeller thank you for your help. I merged your commit and it has passed the tester now. What is the next step?

**Comment by gassmoeller** [comment]
The next step is to press the merge button on my end. Thanks for the contribution!

**Pull request merged**